### ISSUE TITLE
Crash: drcov exits early

### ISSUE THREAD

**Issue by securitykernel** [DynamoRIO/dynamorio#3923]
What version of DynamoRIO are you using?

cronbuild-7.91.18201

Does the latest build from
https://github.com/DynamoRIO/dynamorio/wiki/Latest-Build solve the problem?

No

What operating system version are you running on?

Windows 10 Pro, Version 1903, OS build 18362.418

What application are you running?

custom Visual Studio 2017 generated C++ application

Is your application 32-bit or 64-bit?

32 bit

How are you running the application under DynamoRIO?

bin32\drrun.exe -debug -verbose -t drcov -- application.exe

What happens when you run with debug build ("-debug" flag to
drrun/drconfig/drinject)?

```
INFO: reading tool config file C:\UserProgs\DynamoRIO-Windows/tools/drcov.drrun32
INFO: client 0 path: C:\UserProgs\DynamoRIO-Windows\tools\lib32\release\drcov.dll
INFO: targeting application: "C:\GitLab-Runner\builds\1d533076\0\apps\TrustedWorkstationAgent\Win32\Debug\TrustedWorkstationAgentTests.exe"
INFO: app cmdline:  ".\Win32\Debug\TrustedWorkstationAgentTests.exe"
INFO: configuration directory is "C:\Users\user/dynamorio"
INFO: created child with pid 2052 for C:\GitLab-Runner\builds\1d533076\0\apps\TrustedWorkstationAgent\Win32\Debug\TrustedWorkstationAgentTests.exe
INFO: registering client with id=0 path=|C:\UserProgs\DynamoRIO-Windows\tools\lib32\release\drcov.dll| ops=||
INFO: waiting forever for app to exit...
<Starting application C:\application.exe (2052)>
<Running on newer-than-this-build "Microsoft Windows 10-1903 x64">
<Early threads found>
<Initial options = -no_dynamic_options -client_lib 'C:\UserProgs\DynamoRIO-Windows\tools\lib32\release\drcov.dll;0;' -code_api -probe_api -stack_size 56K -nop_initial_bblock -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -no_aslr_dr -pad_jmps_mark_no_trace >
```

What is the expected output? What do you see instead?  Is this an
application crash, a DynamoRIO crash, a DynamoRIO assert, or a hang (see
https://github.com/DynamoRIO/dynamorio/wiki/Bug-Reporting and set the title
appropriately)?

drrun exits after a few seconds, without running application.exe. The same happens when running any other 32-bit application under drcov, e.g., notepad++.exe.

Please provide any additional information below.

This is a Windows 10 virtual machine. We recently moved it from a Linux server running KVM to a Windows 10 server running Hyper-V. It used to work fine last time it ran under KVM, it is broken since we moved to Hyper-V.

Trying to run 64-bit drcov with a 64 bit application shows the following error:

bin64\\drrun.exe -debug -verbose -t drcov -- C:\windows\notepad.exe

```
Application C:\windows\notepad.exe (16108) DynamoRIO usage error: encode error: rip-relative reference out of 32-bit reach
```

![image](https://user-images.githubusercontent.com/8866185/68212791-8d4b8180-ffda-11e9-9bbe-d9e219c810c5.png)

**Comment by derekbruening** [comment]
Could you attach windbg to notepad.exe and get a callstack while that "encode error" popup is up?  This requires loading symbols via the `load_syms64` script inside windbg: see https://github.com/DynamoRIO/dynamorio/wiki/Debugging#private-libraries

**Comment by securitykernel** [comment]
> Does plain DR (i.e., no -t drcov) work fine for 32-bit and 64-bit?

32 bit exits early:

```
$ bin32\\drrun.exe -debug -verbose -- .\Win32\Debug\TrustedWorkstationAgentTests.exe
INFO: targeting application: "C:\GitLab-Runner\builds\1d533076\0\apps\TrustedWorkstationAgent\Win32\Debug\TrustedWorkstationAgentTests.exe"
INFO: app cmdline:  ".\Win32\Debug\TrustedWorkstationAgentTests.exe"
INFO: configuration directory is "C:\Users\user/dynamorio"
INFO: created child with pid 9100 for C:\GitLab-Runner\builds\1d533076\0\apps\TrustedWorkstationAgent\Win32\Debug\TrustedWorkstationAgentTests.exe
INFO: waiting forever for app to exit...
<Starting application C:\GitLab-Runner\builds\1d533076\0\apps\TrustedWorkstationAgent\Win32\Debug\TrustedWorkstationAgentTests.exe (9100)>
<Running on newer-than-this-build "Microsoft Windows 10-1903 x64">
<Early threads found>
<Initial options = -no_dynamic_options -code_api -probe_api -stack_size 56K -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -no_aslr_dr -pad_jmps_mark_no_trace >
```

64 bit same error message:
No, same error message with `bin64\drrun.exe -debug -verbose -- C:\windows\notepad.exe`

`Application C:\windows\notepad.exe (16108) DynamoRIO usage error: encode error: rip-relative reference out of 32-bit reach`

> Could you attach windbg to notepad.exe and get a callstack while that "encode error" popup is up?

Hm, the callstack doesn't look any useful:

![image](https://user-images.githubusercontent.com/8866185/69038340-e60f1700-09e9-11ea-9797-c6a59db7d599.png)

Maybe the command window does:

![image](https://user-images.githubusercontent.com/8866185/69038403-04751280-09ea-11ea-9c4b-a9016e43138c.png)

**Comment by derekbruening** [comment]
> > Could you attach windbg to notepad.exe and get a callstack while that "encode error" popup is up?
>
> Hm, the callstack doesn't look any useful:

Like I said, please load the symbols to get a useful callstack.  I provided a link there.

**Comment by derekbruening** [comment]
Is this #3949?  The callstack would confirm.

**Comment by securitykernel** [comment]
I did execute the script to load symbols, but it just didn't work.

Unfortunately I get a different error with the cronbuild referenced in #3949:

64 bit:

![image](https://user-images.githubusercontent.com/8866185/69350586-1ff64c80-0c7a-11ea-8162-018437f7a11b.png)

32 bit:

![image](https://user-images.githubusercontent.com/8866185/69350612-2c7aa500-0c7a-11ea-93fb-d4e49cbbc156.png)

**Comment by derekbruening** [comment]
> I did execute the script to load symbols, but it just didn't work.

What command did you run in windbg and what was the output?
It works for me for an assert at that same point in callback_interception_init_finish:
```
$><d:\derek\dr\git\src\tools\windbg-scripts\load_syms64
~0s; kn
```

I can't reproduce the assert on Win10 1903 (18362.476).

```
% DynamoRIO-Windows-7.91.18219-0/bin64/drrun -debug -t drcov -- notepad
<Starting application C:\WINDOWS\system32\notepad.exe (12468)>
<Running on newer-than-this-build "Microsoft Windows 10-1903 x64">
<Early threads found>
<Initial options = -no_dynamic_options -client_lib 'D:\derek\dr\releases\DynamoRIO-Windows-7.91.18219-0\tools\lib64\release\drcov.dll;0;' -code_api -probe_api -stack_size 56K -nop_initial_bblock -max_elide_jmp 0 -max_elide_call 0 -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct -pad_jmps_mark_no_trace >
<curiosity: rex.w on OPSZ_6_irex10_short4!>
<Application changing protections of system memory at least once (0x00007ffe6b99a000-0x00007ffe6b99e000)>
<Stopping application C:\WINDOWS\system32\notepad.exe (12468)>

% ls -t
drcov.notepad.exe.12468.0000.proc.log  ...
```

So it must depend on something particular in your system.

It looks like this has been hit on other systems: https://github.com/DynamoRIO/drmemory/issues/2208

And this is the proposed long-term fix: #2666.  That issue also claims that the problem has been fixed for Skylake.  At this point I am thinking we should close this issue and if you could update #2666 with the context size that you are seeing and what processor you have.

**Comment by derekbruening** [comment]
Closing as suspected duplicate of #3949 and #2666.
